### DISCUSSION TITLE
Github Actions failure handling during deployment

### DISCUSSION THREAD

**Discussion by judoole** [community/community#25900] [Actions]
<p>Given a workflow that is triggered <strong>on = “deployment”</strong>  I would like to both deploy and update the deployment-api. From what I gather I guess these are my alternatives:</p>
<ol>
<li>
<p>Deploy action does a http call to deployments api on failure/success</p>
</li>
<li>
<p>Deploy action saves deploy result in a file and uses exit code 0 for success. The sequential action reads from that file and does a http call to deployments api.</p>
</li>
<li>
<p>Use checks api or a Github App to notify Deployments API on failure/success</p>
</li>
</ol>
<p><strong>1</strong> I don’t like because it is hard to reuse code. <strong>2 </strong> reuses code, but complicates deploy actions.  <strong>3 </strong> would be great, but how?</p>
<p>Anyone doing something similar? All the examples I find, that does deployment, never updates the deployment api. Most of them just does deploy automatically if the branch is the master branch.</p>

**Comment by AndreaGriffiths11** [comment]
<p>Thank you for taking the time to write your feedback! Hopefully, someone else here has a similar workflow and can share their experience. In the meanwhile,  I’ve taken your feedback and passed it along to the appropriate teams. </p>

**Comment by judoole** [comment, marked as answer]
<p>I found a functional solution for this, going along with option number 1.</p>
<p>What I found to be working quite well, was to create a generic action that adds shell scripts which later actions can reuse. The scripts strive to be as minimal as possible, and has only one dependency, and that is <strong>curl</strong>. </p>
<p>It worked pretty well for a couple of projects that we are testing on, and the code is made available at <a href="https://github.com/unacast/actions/tree/master/github-deploy" rel="nofollow noopener">https://github.com/unacast/actions/tree/master/github-deploy</a></p>
<p>I’d much appreciate comments or even better - pull requests :slight_smile:</p>

**Comment by jerrylopez** [comment]
<p>I’ve been trying to come up with a solution for this on and off for the past couple weeks.</p>
<p>Based on my experience I came up with two ideas but it would require help from GitHub.</p>
<ol>
<li>
<p>It seems like it would be great if an action was allowed to run regardless if the action it depended on succeeded or failed. If something like this we’re possible i believe we could have a simple deployment status action that sends the update based on the status of the previous actions, but allowing that would obviously be a flag/option the user enables.</p>
</li>
<li>
<p>Since the workflow is kicked off on deployment events maybe GitHub should fire back the deployment status per action automagically for us. If an action succeeds fire a success action, but if one fails fire off a failure.</p>
</li>
</ol>
<p>What do you think about those? I personally like <span class="hashtag">#2</span> because it just seems cleaner to not have your workflow with a bunch of status actions. Also, in your case your actions don’t have to worry about firing off deployment statuses they just do their job.</p>
<p>I also wonder how we can send a request like that in for consideration?</p>

**Comment by judoole** [comment]
<p><span class="hashtag">#1</span> Yes please :slight_smile:</p>
<p><span class="hashtag">#2</span> That was my initial thought also. But that’s was not how it is implemented to begin with. Maybe later down the line. I think it could be pretty clean. Although it might be difficult for the App that is subscribing to the events from the workflow to extract information that it would like to add to a deployment status event</p>
<blockquote>
<p>I also wonder how we can send a request like that in for consideration?</p>
</blockquote>
<p>I don’t quite know, seems like this is the place. Admins replies to discussions I see, and informs that they are taking the matter to the development team.</p>

**Comment by ThirdEyeRose** [comment]
<p>Hi <a class="mention" href="/jerrylopez">@jerrylopez</a>!</p>
<p>Thanks for the suggestion! While we don’t have an official place to submit feature requests, <a class="mention" href="/judoole">@judoole</a> is correct that GitHub Staff does frequent the GitHub Community Forum and as admins and moderators, we do our best to capture this sort of feedback and share it with the teams who are in charge of designing and improving GitHub.</p>
<p>Speaking of which, I’ve passed your feedback on to the appropriate teams. I can’t promise if we will implement this suggestion or when we can make that happen, but we’re always looking for new solutions and if we do implement this feature, it will be announced in the <a href="https://blog.github.com/changelog/" rel="nofollow noopener">GitHub changelog</a>.</p>
<p>Thanks!</p>

**Comment by assimovt** [comment]
<p>Hey <a class="mention" href="/judoole">@judoole</a> 👋🏽</p>
<p>This looks like a nice and simple solution until GitHub adds a tighter integration with their own Deployment API. Imho, as <a class="mention" href="/jerrylopez">@jerrylopez</a> pointed out, if an action is triggered by a <code>deployment</code> event, it would have be great if GitHub provided an option to update the deployment in question - automatically populating it with target url and description.</p>
<p>In the meanwhile, I’ve looked at your solution and got it working with <a href="https://getslashdeploy.com/" rel="nofollow noopener">SlashDeploy</a> - a chatbot to manage deployments from Slack. I’ve created an <a href="https://github.com/getslashdeploy/actions/tree/master/aws-eb" rel="nofollow noopener">aws-eb</a> custom action, which uses your scripts and now I am finally getting deployment notifications into our Slack :raised_hands:t4:.</p>
<p>There is some room for improvement, so expect from me a PR that would populate log url and deployment description soon :wink:.</p>
<p>Cheers!</p>

**Comment by judoole** [comment]
<p>Thanks <a class="mention" href="/assimovt">@assimovt</a>! And <a href="https://github.com/getslashdeploy/actions/tree/master/aws-eb" rel="nofollow noopener">aws-eb</a> looked great!</p>
<p>I’ve had a quick look at Slashdeploy recently, love the work you are doing! We are considering it to replace Hubot for deployments from Slack. Think I tried to register a couple of weeks ago, but I think it failed, probably because I’m don’t have owner-rights of the Unacast organization or something. Love that you can specify configuration within the repository.</p>
<p>For now, I’ve tested using the <a href="https://github.com/integrations/slack" rel="nofollow noopener">Github Slack integration</a> for notifications. It is trying to add a deploy action in near future, <a href="https://github.com/integrations/slack/issues/688" rel="nofollow noopener">it seems like</a>. But I think it won’t be as feature friendly as Slashdeploy :slight_smile:</p>

**Comment by judoole** [comment]
<p>Btw, did you find a solution to the create a log_url to the action <a class="mention" href="/assimovt">@assimovt</a>? I also asked that in <a href="https://github.community/t/github-actions-url-to-run/16029">https://github.community/t5/GitHub-API-Development-and/Github-Actions-url-to-run/m-p/16471#M506</a></p>

**Comment by assimovt** [comment]
<p>Thx, <a class="mention" href="/judoole">@judoole</a> It’s still in PoC as I can’t get the log URL and description of the deployment. Also, noticed you’ve <a href="https://github.community/t/github-actions-url-to-run/16029">bumped into a similar issue</a> and tested that it’s not being exposed in <code>ENV</code>. I think we have to wait for that to come in from GitHub.</p>
<p>Regarding SlashDeploy registration failure - you should be able to request GitHub app installation from organization/repository admins. Mind dropping me an email to <code>tair&amp;nbsp;[at] getslashdeploy [com]</code> so I could help you out? It’d be really awesome if Unacast benefited from all those <a href="https://docs.getslashdeploy.com/commands" rel="nofollow noopener">features</a> we’ve built to support team deployment workflow :wink:</p>

**Comment by assimovt** [comment]
<p><a class="mention" href="/judoole">@judoole</a> Thanks for open sourcing github-deploy - it’s been really helpful to get our Slack + GitHub Actions + AWS Beanstalk deployments rolling. I’ve blogged about it  👉<a href="https://getslashdeploy.com/blog/slack-deploy-github-actions-aws/" rel="nofollow noopener">https://getslashdeploy.com/blog/slack-deploy-github-actions-aws/</a> check it out!</p>

**Comment by judoole** [comment]
<p>Thanks <a class="mention" href="/assimovt">@assimovt</a>, that was a nice write up as well!</p>
<p>On my todolist is todo a write up of my findings as well. Now I have a great blogpost to reference :) </p>

**Comment by colinjfw** [comment]
<p>I ended up solving a very similar problem to this when building deliverybot (<a href="https://deliverybot.dev" rel="nofollow noopener">https://deliverybot.dev</a>). This is the repo that I’m using to deploy to firebase from actions and set the correct deployment status: <a href="https://github.com/deliverybot/status" rel="nofollow noopener">https://github.com/deliverybot/status</a>.</p>
<p>It doesn’t try and automatically wrap anything just builds on top of <code>if</code> statement functionality in GitHub actions basically. If the previous state was <code>failure()</code> then we push a failure status.</p>

**Comment by judoole** [comment]
<p>Nice <a class="mention" href="/colinjfw">@colinjfw</a> !</p>
<p>We’ve rewritten the status action as well. Decided to use shell over node. Because I don’t particularly like that commiting of the node_modules <a href="https://github.com/unacast/actions-github-deployment-status" rel="nofollow noopener">https://github.com/unacast/actions-github-deployment-status</a>. </p>
